### PR TITLE
Upgrade react scripts to v3

### PR DESCRIPTION
**Pull request by nlaille**
## What changes are proposed in this pull request?
 
Upgrade react-scripts to v3.1.1.

Main purposes : 

- react-scripts v1 has not been updated since August 2018
- enable use of higher version of es syntax, needed for https://github.com/mlflow/mlflow/pull/1556 
 
### Is this a user-facing change? 

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

Not directly for users, but for developers.
 
### What component(s) does this PR affect?
 
- [x] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [x] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by nlaille** [comment]
@Zangr , this one is needed for https://github.com/mlflow/mlflow/pull/1556
Also, I had to disable eslint on a few lines because of https://github.com/eslint/eslint/issues/12117 
And `react-scripts` 3.1.1 requires `babel-eslint` 10.0.2.

File: mlflow/server/js/package.json [hunk 1]
```diff
@@ -83,13 +79,16 @@
...
   },
-  "proxy": {
-    "/ajax-api": {
-      "target": "http://localhost:5000"
-    },
-    "/get-artifact": {
-      "target": "http://localhost:5000",
-      "ws": true
-    }
+  "browserslist": {
```
Comment (Zangr) [inline mlflow/server/js/package.json]: Could you please provide more information on why we need this?

File: mlflow/server/js/src/setupProxy.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+const proxy = require('http-proxy-middleware');
+
+function setupProxyFn(app) {
+  app.use(proxy('/ajax-api', { target: 'http://localhost:5000' }));
+  app.use(proxy('/get-artifact', { target: 'http://localhost:5000', ws: true }));
+}
+
+module.exports = setupProxyFn;
```
Comment (Zangr) [inline mlflow/server/js/src/setupProxy.js]: Where is this used?

**Review by Zangr — CHANGES_REQUESTED** [review]
@nlaille , could you please elaborate more on the impact you mentioned in #1556? Also could you you please provide more general information about this PR in the description section, like why we need to remove `react-app-rewired` and bring in 'react-script'.
> Upgrade to react-scripts v3 has a bit more impact than I thought.

File: mlflow/server/js/src/setupProxy.js (hunk 2 above)
Comment (nlaille) [reply mlflow/server/js/src/setupProxy.js]: It was set in `package.json` before. It is for development purposes.
But in react-script v2 or higher for "complexe" proxy setup, you have to use `setupProxy.js` : https://create-react-app.dev/docs/proxying-api-requests-in-development#configuring-the-proxy-manually 

> Note: You do not need to import this file anywhere. It is automatically registered when you start the development server.

File: mlflow/server/js/package.json (hunk 1 above)
Comment (nlaille) [reply mlflow/server/js/package.json]: I am not sure about why it was not included in the first place. 
But when I tried to build the package with react-scripts v3 it was mandatory.

https://create-react-app.dev/docs/supported-browsers-features#configuring-supported-browsers

**Comment by nlaille** [comment]
@Zangr I add `react-app-rewired` back, I am not sure why I removed it in the first place.
About the impact, it was mainly because of the small changes which are not related to the previous pull request (package.json, setupProxy, ...)

**Comment by smurching** [comment]
@nlaille thanks for looking into this! We actually tried to react-scripts v3 a few months ago in https://github.com/mlflow/mlflow/pull/968, but found that it broke the search UI, in that it became impossible to enter input into the "filter params", "filter metrics", or "search" fields. Unfortunately, the same still appears to be true - would you be able to look into this?

**Comment by nlaille** [comment]
@smurching UI breaks come from : `ExperimentView:getDerivedStateFromProps` which computes state from props of ExperimentView.

According to some french documentations (I was not able to find it in english) : https://fr.reactjs.org/docs/react-component.html#the-component-lifecycle 

`getDerivedStateFromProps` is called at the creation of a component which is perfect for our use case to get inputs/filters from urls or persistedState. Inputs/filters is injected to ExperimentView using props from ExperimentPage (https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/components/ExperimentPage.js#L250)
```
return <ExperimentView
      paramKeyFilter={new KeyFilter(this.state.persistedState.paramKeyFilterString)}
      metricKeyFilter={new KeyFilter(this.state.persistedState.metricKeyFilterString)}
      experimentId={this.props.experimentId}
      searchRunsRequestId={this.searchRunsRequestId}
      lifecycleFilter={this.state.lifecycleFilter}
      onSearch={this.onSearch}
      searchRunsError={searchRunsError}
      searchInput={this.state.persistedState.searchInput}
      isLoading={isLoading && !searchRunsError}
      orderByKey={this.state.persistedState.orderByKey}
      orderByAsc={this.state.persistedState.orderByAsc}
      nextPageToken={this.state.nextPageToken}
      handleLoadMoreRuns={this.handleLoadMoreRuns}
      loadingMore={this.state.loadingMore}
    />;
```

But, it is also called during component updates, and this part is breaking the UI because in  `ExperimentView:getDerivedStateFromProps` we are get inputs/filters from props instead of the prevState. Because of the first statement.
https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/components/ExperimentView.js#L194 

I have not check if this lifecycle has changed between react versions. But I have removed inputs/filters from `ExperimentView:getDerivedStateFromProps`, and updates are working. But it breaks the behavior at component creation.

I will keep investigate a bit about previous version behavior on `getDerivedStateFromProps`.

**Comment by smurching** [comment]
@nlaille interesting, it is unfortunate that the problem seems to be react lifecycle methods not behaving as expected (seems hard to work around) - I'll take a look. I guess as you found in https://github.com/octokit/rest.js/issues/828, it's tricky to avoid upgrading create-react-app altogether as we need to upgrade to a more modern version to use packages with ES6 syntax.

**Comment by smurching** [comment]
@nlaille thanks for investigating, your analysis on the behavior of `getDerivedStateFromProps` seems correct (basically, we were misusing it before). I opened https://github.com/mlflow/mlflow/pull/1869 to correct that to unblock this PR (your PR plus the changes in https://github.com/mlflow/mlflow/pull/1869 seems to result in a working search UI :))

**Comment by nlaille** [comment]
@smurching there is some warnings from UNSAFE_methods, and some outdated calls from old version of some packages like `react-router-*`, `react-redux`. 

Do you mind if I upgrade them as well ? Or should it be another pr ?

**Pull request closed without merging**